design-documents/jwt-support.md

```diff
@@ -0,0 +1,180 @@
+
+#### Data exchange
+
+![Data exchange](img/jwt-data-exchange.png)
+
+Allows to built communication between multiple servers. An authorization server shares secret keys between servers or servers can send JWT to the authorization server for the future verification.
+
+#### Data verification
+
+![Data verification](img/jwt-data-verification.png)
+
+Allows to verify if data is received from trusted source. The diagram shows RSA keys usage for content encryption but different types of keys can be used for the data verification like octet strings, key files, X.509 Certificates, etc.
+
+The JWT structure has three main parts:
+
+ - `header` - contains information about signature verification algorithms
+ - `payload` - contains data (JWT-claims), the RFC defines a list of standard optional claims (https://tools.ietf.org/html/rfc7519#section-4.1)
+ - `signature` - is used for data verification and represented as a hash of encoded `header` and encoded payload and secret key
+
+And in general JWT looks like this: `header`.`payload`.`signature`.
+
+The JWT usage became more and more popular for authentication and data verification purposes and Magento has a lot of integrations with different 3rd party systens, we should support JWT creation/validation/parsing out-of-box. Another benefit, that JWT is language agnostic and token generaten with one programming language can be parsed with another, the only key should be shared between applications.
+
+## Solution
+
+We do not need to implement own solution for key generation, parsing tokens, encryption/decryption, etc. [There are](https://jwt.io/#libraries) multiple PHP libraries which support all needed operations and algorithms and we need to create own wrappers.
+
+[PHP JWT Framework](https://github.com/web-token/jwt-framework) - is proposed as a library because:
+ - implements all algorithms from RFC
+ - provides implementation for JWS, JWT, JWE, JWA, JWK, JSON Web Key Thumbprint, Unencoded Payload Option
+ - supports different serialization modes
+ - supports multiple compression methods
+ - full support of JSON Web Key Set
+ - has a good documentation
+ - supports detached payload, multiple signatures, nested tokens
+
+## Implementation
+
+The following diagram represents needed interfaces to unify the workflow for JWT usage.
+
+![Class diagram](img/jwt-class-diagram.png)
+
+The `\Magento\Framework\Jwt\KeyGeneratorInterface` will provide a possibility to use different types of key generators like: Magento deployment secret key, X.509 certificates or RSA keys.
+```php
+interface KeyGeneratorInterface
+{
+    public function create(): JWK;
+}
+
+class CryptKeyGenerator implements KeyGeneratorInterface
+{
+    public function __construct(AlgorithmFactory $algorithmFactory, DeploymentConfig $deploymentConfig)
+    {
+        $this->algorithmFactory = $algorithmFactory;
+        $this->deploymentConfig = $deploymentConfig;
+    }
+
+    public function create(): JWK
```

Comment:
I think it would be better if default key generator would utilize EncryptorInterface - we always update it to follow the most recent security requirements

Reply:
EncryptorInterface doesn't allow to work with 3rd party systems because they can use the different algorithms to generate the key (which are not supported by EncryptorInterface).

Reply:
Like I've said, I would like to see it being used by default signature generator only. Developers should be able to configure other algorithms